File: typescript/strain/strain.ts
type Predicate<T> = (n: T) => Boolean

function not<T>(p: Predicate<T>): Predicate<T> {
    return (n: T) => !p(n)
}

export function keep<T>(collection: T[], predicate: Predicate<T>): T[] {
    const result: T[] = []

    collection.forEach((value) => {
        if (predicate(value)) { result.push(value) }
    })

    return result
}

export function discard<T>(collection: T[], predicate: Predicate<T>): T[] {
    return keep(collection, not(predicate))
}

File: scala/strain/src/main/scala/Strain.scala
object Strain {
  def keep[A](list: Iterable[A], predicate: A => Boolean): Iterable[A] = for (item <- list if predicate(item)) yield item
  def discard[A](list: Iterable[A], predicate: A => Boolean): Iterable[A] = keep[A](list, v => !predicate(v))
}